File: mobile/app/src/main/java/dam/isi/frsf/utn/edu/ar/delivery/service/DataService.java
package dam.isi.frsf.utn.edu.ar.delivery.service;

import android.content.Context;

import com.google.gson.reflect.TypeToken;
import com.koushikdutta.ion.Ion;
import com.koushikdutta.ion.future.ResponseFuture;

import java.util.List;

import dam.isi.frsf.utn.edu.ar.delivery.model.Topping;
import dam.isi.frsf.utn.edu.ar.delivery.model.ContainerType;
import dam.isi.frsf.utn.edu.ar.delivery.model.Deal;
import dam.isi.frsf.utn.edu.ar.delivery.model.Flavor;
import dam.isi.frsf.utn.edu.ar.delivery.model.Location;
import dam.isi.frsf.utn.edu.ar.delivery.model.Order;
import dam.isi.frsf.utn.edu.ar.delivery.model.Review;
import dam.isi.frsf.utn.edu.ar.delivery.model.Sauce;

import static dam.isi.frsf.utn.edu.ar.delivery.constants.appConstants.deviceID;
import static dam.isi.frsf.utn.edu.ar.delivery.constants.appConstants.localPath;

public class DataService {

    private Context context;

    public DataService(Context context) {
        this.context = context;
    }

    public ResponseFuture<Deal> getLastDeal() throws Exception {
        String path = localPath + "deals";
        return Ion.with(context).load(path).as(new TypeToken<Deal>() {
        });
    }

    public ResponseFuture<List<Flavor>> getFlavors() throws Exception {
        String path = localPath + "flavors";
        return Ion.with(context).load(path).as(new TypeToken<List<Flavor>>() {
        });
    }

    public ResponseFuture<List<ContainerType>> getContainers() throws Exception {
        String path = localPath + "containers";
        return Ion.with(context).load(path).as(new TypeToken<List<ContainerType>>() {
        });
    }

    public ResponseFuture<List<Topping>> getToppings() throws Exception {
        String path = localPath + "toppings";
        return Ion.with(context).load(path).as(new TypeToken<List<Topping>>() {});
    }

    public ResponseFuture<List<Sauce>> getSauces() throws Exception {
        String path = localPath + "sauces";
        return Ion.with(context).load(path).as(new TypeToken<List<Sauce>>() {});
    }

    public ResponseFuture<String> addOrder(Order order) throws Exception {
        String path = localPath + "session/" + deviceID + "/order";
        return Ion.with(context).load(path).setJsonPojoBody(order).as(new TypeToken<String>() {
        });
    }

    public ResponseFuture<String> modifyOrder(String orderID, Order order) throws Exception {
        String path = localPath + "session/" + deviceID + "/order/" + orderID;
        return Ion.with(context).load(path).setJsonPojoBody(order).as(new TypeToken<String>() {
        });
    }

    public ResponseFuture<List<Order>> getOrders() throws Exception {
        String path = localPath + "session/" + deviceID + "/orders";
        return Ion.with(context).load(path).as(new TypeToken<List<Order>>() {
        });
    }

    public ResponseFuture<Location> getlocation(String orderID) throws Exception {
        String path = localPath + "location/" + orderID;
        return Ion.with(context).load(path).as(new TypeToken<Location>() {
        });
    }

    public ResponseFuture<String> addReview(Review review) throws Exception {
        String path = localPath + "review";
        return Ion.with(context).load(path).setJsonPojoBody(review).as(new TypeToken<String>() {
        });
    }

}

File: mobile/app/src/main/java/dam/isi/frsf/utn/edu/ar/delivery/service/DataService.kt
package dam.isi.frsf.utn.edu.ar.delivery.service

import android.content.Context
import com.google.gson.reflect.TypeToken
import com.koushikdutta.ion.Ion
import com.koushikdutta.ion.future.ResponseFuture
import dam.isi.frsf.utn.edu.ar.delivery.constants.appConstants.deviceID
import dam.isi.frsf.utn.edu.ar.delivery.constants.appConstants.localPath
import dam.isi.frsf.utn.edu.ar.delivery.model.*

class DataService(private val context: Context) {

    val lastDeal: ResponseFuture<List<Deal>>
        @Throws(Exception::class)
        get() {
            val path = localPath!! + "deals"
            return Ion.with(context).load(path).`as`(object : TypeToken<List<Deal>>() {
            })
        }

    val flavors: ResponseFuture<List<Flavor>>
        @Throws(Exception::class)
        get() {
            val path = localPath!! + "flavors"
            return Ion.with(context).load(path).`as`(object : TypeToken<List<Flavor>>() {

            })
        }

    val containers: ResponseFuture<List<ContainerType>>
        @Throws(Exception::class)
        get() {
            val path = localPath!! + "containers"
            return Ion.with(context).load(path).`as`(object : TypeToken<List<ContainerType>>() {

            })
        }

    val toppings: ResponseFuture<List<Topping>>
        @Throws(Exception::class)
        get() {
            val path = localPath!! + "toppings"
            return Ion.with(context).load(path).`as`(object : TypeToken<List<Topping>>() {

            })
        }

    val sauces: ResponseFuture<List<Sauce>>
        @Throws(Exception::class)
        get() {
            val path = localPath!! + "sauces"
            return Ion.with(context).load(path).`as`(object : TypeToken<List<Sauce>>() {

            })
        }

    @Throws(Exception::class)
    fun addOrder(order: Order): ResponseFuture<String> {
        val path = localPath + "session/" + deviceID + "/order"
        return Ion.with(context).load(path).setJsonPojoBody(order).`as`(object : TypeToken<String>() {

        })
    }

    @Throws(Exception::class)
    fun modifyOrder(orderID: String, order: Order): ResponseFuture<String> {
        val path = localPath + "session/" + deviceID + "/order/" + orderID
        return Ion.with(context).load(path).setJsonPojoBody(order).`as`(object : TypeToken<String>() {

        })
    }

    val orders: ResponseFuture<List<Order>>
        @Throws(Exception::class)
        get() {
            val path = localPath + "session/" + deviceID + "/orders"
            return Ion.with(context).load(path).`as`(object : TypeToken<List<Order>>() {

            })
        }

    @Throws(Exception::class)
    fun getlocation(orderID: String): ResponseFuture<Location> {
        val path = localPath + "location/" + orderID
        return Ion.with(context).load(path).`as`(object : TypeToken<Location>() {

        })
    }

    @Throws(Exception::class)
    fun addReview(review: Review): ResponseFuture<String> {
        val path = localPath!! + "review"
        return Ion.with(context).load(path).setJsonPojoBody(review).`as`(object : TypeToken<String>() {

        })
    }

}
